Source/CarthageKit/Project.swift

```diff
@@ -733,20 +742,46 @@ public final class Project {
 
 	/// Attempts to build each Carthage dependency that has been checked out,
 	/// optionally they are limited by the given list of dependency names.
+	/// Cached dependencies whose dependency trees are also cached will not
+	/// be rebuilt unless otherwise specified via build options.
 	///
 	/// Returns a producer-of-producers representing each scheme being built.
 	public func buildCheckedOutDependenciesWithOptions(_ options: BuildOptions, dependenciesToBuild: [String]? = nil, sdkFilter: @escaping SDKFilterCallback = { .success($0.0) }) -> SignalProducer<BuildSchemeProducer, CarthageError> {
 		return loadResolvedCartfile()
 			.flatMap(.merge) { resolvedCartfile in
 				return self.buildOrderForResolvedCartfile(resolvedCartfile, dependenciesToInclude: dependenciesToBuild)
+					.flatMap(.concat) { dependency -> SignalProducer<(Dependency<PinnedVersion>, Set<ProjectIdentifier>), CarthageError> in
+						self.dependencyProjects(for: dependency)
+							.map { projects in (dependency, projects) }
+					}
+					.reduce([]) { (includedDependencies, nextGroup) -> [Dependency<PinnedVersion>] in
+						let (nextDependency, projects) = nextGroup
+						guard options.cacheBuilds else {
+							return includedDependencies + [nextDependency]
+						}
+
+						var dependenciesToBuild = includedDependencies
+						_ = self.shouldSkipBuildForDependency(nextDependency, dependencyProjects: projects, dependenciesToBeBuilt: includedDependencies, platforms: options.platforms)
+							.on(value: { shouldSkip in
+								if !shouldSkip {
+									dependenciesToBuild.append(nextDependency)
+								}
+							})
+							.wait() // We must wait since subsequent calls to reduce rely on the results of the previous project skipping or not
```

Comment:
I think it should be possible to do this without a `wait` if you move this check:

```swift
		let projectsToBeBuilt = Set(dependenciesToBeBuilt.map { $0.project })
		guard dependencyProjects.intersection(projectsToBeBuilt).isEmpty else {
			return .init(value: false)
		}
```

1. Get build order
2. Get dependencies for project
3. Get whether project has cached version
4. Reduce `(Dependency, Set<ProjectIdentifier>, Bool)` into a set of projects to be built